File: core/java/src/main/java/hara/lang/protocol/IReduce.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Reduces a value with or without an explicit initial accumulator. */
@HaraProtocolBinding(namespace = "std.protocol.ireduce", name = "IReduce")
public interface IReduce {
  Object reduce(Object function);

  @HaraMethod(value = "reduce", arity = -1, variadic = true)
  Object reduce(Object function, Object initial);
}

File: core/rust/src/lang/protocol/ireduce.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ireduce", name = "IReduce")]
pub trait IReduce<A, F> {
    type Error;

    #[hara_method(value = "reduce", arity = -1, variadic = true, min_arity = 2, max_arity = 3)]
    fn reduce(&self, function: F, initial: Option<A>) -> Result<A, Self::Error>;
}
